src/coreclr/src/vm/prestub.cpp

```diff
@@ -1767,113 +1767,155 @@ extern "C" MethodDesc * STDCALL PreStubGetMethodDescForCompactEntryPoint (PCODE
+{
+    _ASSERTE(pMD->HasNativeCallableAttribute());
+    _ASSERTE(pMD->IsStatic() // Must be static
+             && !pMD->IsGenericMethodDefinition() // No generics
+             && !NDirect::MarshalingRequired(pMD, pMD->GetSig(), pMD->GetModule())); // Blittable arguments
+
+    PCODE pbRetVal = NULL;
+
+    STATIC_CONTRACT_THROWS;
+    STATIC_CONTRACT_GC_TRIGGERS;
+    STATIC_CONTRACT_MODE_PREEMPTIVE;
+
+    MAKE_CURRENT_THREAD_AVAILABLE();
+
+    // No GC frame is needed here since there should be no OBJECTREFs involved
+    // in this call due to NativeCallableAttribute semantics.
+
+    INSTALL_MANAGED_EXCEPTION_DISPATCHER;
+    INSTALL_UNWIND_AND_CONTINUE_HANDLER;
+
+    // Make sure the method table is restored, and method instantiation if present
+    pMD->CheckRestore();
+    CONSISTENCY_CHECK(GetAppDomain()->CheckCanExecuteManagedCode(pMD));
+
+    pbRetVal = pMD->DoPrestub(NULL);
+
+    UNINSTALL_UNWIND_AND_CONTINUE_HANDLER;
+    UNINSTALL_MANAGED_EXCEPTION_DISPATCHER;
+
+    {
+        HardwareExceptionHolder;
+
+        // Give debugger opportunity to stop here
+        ThePreStubPatch();
+    }
+
+    return pbRetVal;
+}
+
 //=============================================================================
 // This function generates the real code for a method and installs it into
 // the methoddesc. Usually ***BUT NOT ALWAYS***, this function runs only once
 // per methoddesc. In addition to installing the new code, this function
 // returns a pointer to the new code for the prestub's convenience.
 //=============================================================================
-extern "C" PCODE STDCALL PreStubWorker(TransitionBlock * pTransitionBlock, MethodDesc * pMD)
+extern "C" PCODE STDCALL PreStubWorker(TransitionBlock* pTransitionBlock, MethodDesc* pMD)
 {
     PCODE pbRetVal = NULL;
 
     BEGIN_PRESERVE_LAST_ERROR;
 
     STATIC_CONTRACT_THROWS;
     STATIC_CONTRACT_GC_TRIGGERS;
-    STATIC_CONTRACT_MODE_COOPERATIVE;
+    STATIC_CONTRACT_MODE_ANY;
     STATIC_CONTRACT_ENTRY_POINT;
 
-    MAKE_CURRENT_THREAD_AVAILABLE();
-
-#ifdef _DEBUG
-    Thread::ObjectRefFlush(CURRENT_THREAD);
-#endif
-
-    FrameWithCookie<PrestubMethodFrame> frame(pTransitionBlock, pMD);
-    PrestubMethodFrame * pPFrame = &frame;
-
-    pPFrame->Push(CURRENT_THREAD);
-
-    INSTALL_MANAGED_EXCEPTION_DISPATCHER;
-    INSTALL_UNWIND_AND_CONTINUE_HANDLER;
+    _ASSERTE(!NingenEnabled() && "You cannot invoke managed code inside the ngen compilation process.");
 
-    ETWOnStartup (PrestubWorker_V1,PrestubWorkerEnd_V1);
+    ETWOnStartup(PrestubWorker_V1, PrestubWorkerEnd_V1);
 
-    _ASSERTE(!NingenEnabled() && "You cannot invoke managed code inside the ngen compilation process.");
+    MAKE_CURRENT_THREAD_AVAILABLE();
 
-    // Running the PreStubWorker on a method causes us to access its MethodTable
-    g_IBCLogger.LogMethodDescAccess(pMD);
+    // Check what GC mode we are running under.
+    if (!CURRENT_THREAD->PreemptiveGCDisabled())
```

Comment:
`CURRENT_THREAD` can be null here

Reply:
Hmmm. Then that is a long standing bug. The `CURRENT_THREAD` was used in the `pPFrame->Push(CURRENT_THREAD);` from the original code and that dereferences the thread in https://github.com/dotnet/runtime/blob/cf4cc83d44f3ebd1807b6db0ad9e546bdce218b7/src/coreclr/src/vm/frames.cpp#L386-L398

Should I convert this to the same approach as `JIT_ReversePInvokeEnter`? Which I guess at this point would mean the thread would be created here and that call could potentially be asserted instead.

Reply:
It can be NULL only for the reverse PInvoke case that you are introducing. I would change this condition to `if (CURRENT_THREAD == NULL || !CURRENT_THREAD->PreemptiveGCDisabled())` and deal with creating the thread in `PreStubWorker_Preemptive`.

Reply:
Oh right. The PREEMPT mode is new... duh. That makes sense.